In [3]:
import pandas as pd
import numpy as np
import os



In [4]:
# load data from a file

df = pd.read_csv("aqi.csv")

In [5]:
# Overview of structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10116546 entries, 0 to 10116545
Data columns (total 10 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   _id                     int64  
 1   datetime_est            object 
 2   site                    object 
 3   parameter               object 
 4   is_valid                object 
 5   report_value            float64
 6   unit                    object 
 7   unit_description        object 
 8   highest_flag            object 
 9   aqs_parameter_category  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 771.8+ MB


In [6]:

# Look at the first few rows
df.head()

_id         datetime_est     site parameter is_valid  report_value  \
0    1  2016-01-01T00:00:00  Lincoln      PM25        t         3.000   
1    2  2016-01-01T00:00:00  Lincoln    PER_F2        t        67.000   
2    3  2016-01-01T00:00:00  Liberty       SO2        t         0.001   
3    4  2016-01-01T00:00:00  Liberty   SONICWS        t         6.900   
4    5  2016-01-01T00:00:00  Liberty   SONICWD        t       259.000   

      unit            unit_description highest_flag aqs_parameter_category  
0    UG/M3  micrograms per cubic meter          NaN            Particulate  
1  PERCENT                     Percent          NaN                    NaN  
2      PPM           Parts Per Million            C    Criteria Pollutants  
3      MPH              miles per hour          NaN         Meteorological  
4      DEG                     Degrees          NaN         Meteorological

In [7]:
# Descriptive statistics for numeric columns
df.describe()

_id  report_value
count  1.011655e+07  8.537712e+06
mean   5.075763e+06  5.134890e+02
std    2.938486e+06  5.002255e+03
min    1.000000e+00 -2.034370e+04
25%    2.529137e+06  2.400000e+00
50%    5.058274e+06  1.670000e+01
75%    7.631409e+06  4.790000e+01
max    1.016082e+07  1.002960e+05

In [8]:
# Check for missing values
df.isnull().sum()

_id                             0
datetime_est                    0
site                            0
parameter                       0
is_valid                        0
report_value              1578834
unit                       215109
unit_description             2894
highest_flag              8322242
aqs_parameter_category    4884545
dtype: int64

In [9]:

# Check column types
df.dtypes

_id                         int64
datetime_est               object
site                       object
parameter                  object
is_valid                   object
report_value              float64
unit                       object
unit_description           object
highest_flag               object
aqs_parameter_category     object
dtype: object

In [10]:
df.columns.value_counts()

_id                       1
datetime_est              1
site                      1
parameter                 1
is_valid                  1
report_value              1
unit                      1
unit_description          1
highest_flag              1
aqs_parameter_category    1
Name: count, dtype: int64

In [11]:
# Check for duplicates
df.duplicated().sum()

0

In [12]:
# Quick look at categorical distributions
df.select_dtypes(include='object').apply(pd.Series.value_counts)

datetime_est  site  parameter   is_valid  unit  \
+                                       NaN   NaN        NaN        NaN   NaN   
-                                       NaN   NaN        NaN        NaN   NaN   
2016-01-01T00:00:00                   104.0   NaN        NaN        NaN   NaN   
2016-01-01T01:00:00                   104.0   NaN        NaN        NaN   NaN   
2016-01-01T02:00:00                   104.0   NaN        NaN        NaN   NaN   
...                                     ...   ...        ...        ...   ...   
millimeters                             NaN   NaN        NaN        NaN   NaN   
millimeters of mercury (torr)           NaN   NaN        NaN        NaN   NaN   
nanograms per cubic meter               NaN   NaN        NaN        NaN   NaN   
t                                       NaN   NaN        NaN  8537712.0   NaN   
watts per square meter                  NaN   NaN        NaN        NaN   NaN   

                               unit_description  highest_flag  \
+                                           NaN        9624.0   
-                                           NaN       45966.0   
2016-01-01T00:00:00                         NaN           NaN   
2016-01-01T01:00:00                         NaN           NaN   
2016-01-01T02:00:00                         NaN           NaN   
...                                         ...           ...   
millimeters                             49462.0           NaN   
millimeters of mercury (torr)          419446.0           NaN   
nanograms per cubic meter              162720.0           NaN   
t                                           NaN           NaN   
watts per square meter                 133656.0           NaN   

                               aqs_parameter_category  
+                                                 NaN  
-                                                 NaN  
2016-01-01T00:00:00                               NaN  
2016-01-01T01:00:00                               NaN  
2016-01-01T02:00:00                               NaN  
...                                               ...  
millimeters                                       NaN  
millimeters of mercury (torr)                     NaN  
nanograms per cubic meter                         NaN  
t                                                 NaN  
watts per square meter                            NaN  

[81576 rows x 8 columns]

Clean datetime_est column. Change it do datetime format

In [13]:
# Convert to datetime 
df['datetime_est'] = pd.to_datetime(df['datetime_est'])

# Check the min and max timestamps
df['datetime_est'].min(), df['datetime_est'].max()

(Timestamp('2016-01-01 00:00:00'), Timestamp('2025-04-13 08:00:00'))

In [14]:
df['site'].nunique()

20

In [15]:
# get the unique values in the site column
df['site'].unique()

array(['Lincoln', 'Liberty', 'North Braddock', 'Lawrenceville 2',
       'Parkway East', 'Liberty 2', 'Harrison Township', 'Lawrenceville',
       'South Fayette', 'Avalon', 'Flag Plaza', 'Court House',
       'Glassport High Street', 'Opal Court', 'Clairton', 'West Mifflin',
       'Pittsburgh', 'USS Clairton Sodar Site', 'Fulton St. Fridge',
       'Lawrencevill00:00'], dtype=object)

Explore and find out what 'Lawrencevill00:00' is

In [16]:
# load in the site data

site_df = pd.read_csv("sensor locations.csv")




In [17]:
# explore the site data
site_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   _id               25 non-null     int64  
 1   site_name         25 non-null     object 
 2   description       22 non-null     object 
 3   air_now_mnemonic  5 non-null      object 
 4   address           20 non-null     object 
 5   latitude          21 non-null     float64
 6   longitude         21 non-null     float64
 7   enabled           25 non-null     object 
dtypes: float64(2), int64(1), object(5)
memory usage: 1.7+ KB


In [18]:
site_df.head()

_id          site_name                                       description  \
0  4916  Fulton St. Fridge          Sample Fridge at 836 Fulton St. Room 117   
1  4917  Harrison Township           Harrison Township Ozone and NO/NOx Site   
2  4918      South Fayette                          South Fayette Ozone Site   
3  4919      Lawrenceville                 Lawrenceville Ozone, and Met Site   
4  4920            Liberty  Liberty Borough South Allegheny High School Site   

  air_now_mnemonic                                            address  \
0              NaN             836 Fulton Street Pittsburgh, PA 15233   
1              HT1  Highlands Senior High School - 1500 Pacific Av...   
2              SF1  South Fayette Elementary School - 3640 Old Oak...   
3              LV1  Allegheny County Health Department - 301 39th ...   
4              NaN  South Allegheny High School - 2743 Washington ...   

    latitude  longitude enabled  
0        NaN        NaN       t  
1  40.613864 -79.729569       t  
2  40.375647 -80.169931       t  
3  40.465433 -79.960742       t  
4  40.323856 -79.868064       t

In [20]:
site_df.site_name.unique()

array(['Fulton St. Fridge', 'Harrison Township', 'South Fayette',
       'Lawrenceville', 'Liberty', 'West Allegheny', 'Avalon',
       'Lawrenceville 2', 'Monroeville', 'North Braddock', 'Clairton',
       'Pittsburgh 8_Manchester', 'Lincoln', 'Parkway East',
       'Natrona Lead', 'Bridgeville', 'Liberty 2',
       'Glassport High Street', 'Flag Plaza', 'Court House', 'Opal Court',
       'West Mifflin', 'USS Clairton Sodar Site',
       '836 Fulton Street Office', 'Liberty Trailer'], dtype=object)

In [21]:
# Unique site names in each DataFrame
main_sites = set(df['site'].unique())
reference_sites = set(site_df['site_name'].unique())

# Find sites in df that don't exist in site_df
unmatched_sites = main_sites - reference_sites
print(unmatched_sites)

{'Pittsburgh', 'Lawrencevill00:00'}


In [22]:
unmatched_sites_2 = reference_sites - main_sites
print(unmatched_sites_2)

{'Bridgeville', 'West Allegheny', 'Pittsburgh 8_Manchester', 'Natrona Lead', 'Liberty Trailer', '836 Fulton Street Office', 'Monroeville'}


The following locations are in the site locations data file but not in the main pollution file.

'Bridgeville', 'West Allegheny', 'Pittsburgh 8_Manchester', 'Natrona Lead', 'Liberty Trailer', '836 Fulton Street Office', 'Monroeville'

The following locations are in the main pollution file but not in the site locations data file.

'Pittsburgh', 'Lawrencevill00:00'

I infer that Pittsburgh 8_Manchester and Pittsburgh are the same.
'Bridgeville', 'West Allegheny', 'Natrona Lead', 'Liberty Trailer', '836 Fulton Street Office', and 'Monroeville' have no records in the main data file

I still need to find out about 'Lawrencevill00:00' and see what it means

In [23]:
# Check for records with location 'Lawrencevill00:00'
df['site'].value_counts().loc['Lawrencevill00:00']

1

In [24]:
# Identify the rows with 'Lawrencevill00:00'
df[df['site'] == 'Lawrencevill00:00']

_id        datetime_est               site parameter is_valid  \
9426564  9470741 2024-10-04 05:00:00  Lawrencevill00:00       6.2        f   

         report_value  unit unit_description highest_flag  \
9426564           NaN  TRUE              NaN            <   

        aqs_parameter_category  
9426564             Pittsburgh

This entry looks like like an error. We will go ahead to delete that entry

In [25]:
# delete the rows with 'Lawrencevill00:00'
df = df[df['site'] != 'Lawrencevill00:00']

In [28]:
# Identify 'Lawrenceville 2' site 
site_df[site_df['site_name'] == 'Lawrenceville 2']

_id        site_name     description air_now_mnemonic  \
7  4923  Lawrenceville 2  LV2 NCORE site              NaN   

                                             address   latitude  longitude  \
7  Allegheny County Health Department - 01 39th S...  40.465433 -79.960742   

  enabled  
7       t

In [29]:
# Identify 'Lawrenceville' site 
site_df[site_df['site_name'] == 'Lawrenceville']

_id      site_name                        description air_now_mnemonic  \
3  4919  Lawrenceville  Lawrenceville Ozone, and Met Site              LV1   

                                             address   latitude  longitude  \
3  Allegheny County Health Department - 301 39th ...  40.465433 -79.960742   

  enabled  
3       t

From a quick google maps search, the locations are different from each other and does not appear to be an error.
There are two sensor devices in Lawrenceville.

In [31]:
# Decide to encode the site names as categorical variables or one- hot for the model

In [32]:
# convert the site names to categorical variables
df['site'] = df['site'].astype('category')
